### PR TITLE
Update linux images to include .NET Core 3.1

### PR DESCRIPTION
**Pull request by MichaelSimons**
With this change, the following .NET Core components will get installed.

```
 dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.1.100
 Commit:    cd82f021f4

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  18.04
 OS Platform: Linux
 RID:         ubuntu.18.04-x64
 Base Path:   /usr/share/dotnet/sdk/3.1.100/

Host (useful for support):
  Version: 3.1.0
  Commit:  65f04fb6db

.NET Core SDKs installed:
  2.1.300 [/usr/share/dotnet/sdk]
  2.1.301 [/usr/share/dotnet/sdk]
  2.1.302 [/usr/share/dotnet/sdk]
  2.1.401 [/usr/share/dotnet/sdk]
  2.1.402 [/usr/share/dotnet/sdk]
  2.1.403 [/usr/share/dotnet/sdk]
  2.1.500 [/usr/share/dotnet/sdk]
  2.1.502 [/usr/share/dotnet/sdk]
  2.1.503 [/usr/share/dotnet/sdk]
  2.1.504 [/usr/share/dotnet/sdk]
  2.1.505 [/usr/share/dotnet/sdk]
  2.1.506 [/usr/share/dotnet/sdk]
  2.1.507 [/usr/share/dotnet/sdk]
  2.1.508 [/usr/share/dotnet/sdk]
  2.1.509 [/usr/share/dotnet/sdk]
  2.1.510 [/usr/share/dotnet/sdk]
  2.1.602 [/usr/share/dotnet/sdk]
  2.1.603 [/usr/share/dotnet/sdk]
  2.1.604 [/usr/share/dotnet/sdk]
  2.1.605 [/usr/share/dotnet/sdk]
  2.1.606 [/usr/share/dotnet/sdk]
  2.1.607 [/usr/share/dotnet/sdk]
  2.1.700 [/usr/share/dotnet/sdk]
  2.1.701 [/usr/share/dotnet/sdk]
  2.1.801 [/usr/share/dotnet/sdk]
  2.1.802 [/usr/share/dotnet/sdk]
  2.2.100 [/usr/share/dotnet/sdk]
  2.2.101 [/usr/share/dotnet/sdk]
  2.2.102 [/usr/share/dotnet/sdk]
  2.2.103 [/usr/share/dotnet/sdk]
  2.2.104 [/usr/share/dotnet/sdk]
  2.2.105 [/usr/share/dotnet/sdk]
  2.2.106 [/usr/share/dotnet/sdk]
  2.2.107 [/usr/share/dotnet/sdk]
  2.2.108 [/usr/share/dotnet/sdk]
  2.2.109 [/usr/share/dotnet/sdk]
  2.2.110 [/usr/share/dotnet/sdk]
  2.2.202 [/usr/share/dotnet/sdk]
  2.2.203 [/usr/share/dotnet/sdk]
  2.2.204 [/usr/share/dotnet/sdk]
  2.2.205 [/usr/share/dotnet/sdk]
  2.2.206 [/usr/share/dotnet/sdk]
  2.2.207 [/usr/share/dotnet/sdk]
  2.2.300 [/usr/share/dotnet/sdk]
  2.2.301 [/usr/share/dotnet/sdk]
  2.2.401 [/usr/share/dotnet/sdk]
  2.2.402 [/usr/share/dotnet/sdk]
  3.0.100 [/usr/share/dotnet/sdk]
  3.0.101 [/usr/share/dotnet/sdk]
  3.1.100 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.1 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.4 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.6 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.7 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.8 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.9 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.10 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.11 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.12 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.13 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.14 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.1 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.4 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.6 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.7 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.8 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.1 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.4 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.6 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.7 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.8 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.9 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.10 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.11 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.12 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.13 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.14 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.1 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.2 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.4 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.5 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.6 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.7 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.8 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.0.1 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.1 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.6 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.7 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.8 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.9 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.10 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.11 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.12 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.13 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.14 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.1 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.2 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.4 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.5 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.6 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.7 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.8 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.0.1 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
```

### CODE REVIEW

**Comment by RobertDougan** [comment]
Hi @MichaelSimons , is there a timescale for when this will be merged and released?

**Comment by MichaelSimons** [comment]
Hi @RobertDougan - IDK, I am waiting for a repo owner to review the PR and provide feedback.

@alepauly, can you tell me who could review this change?  It is something we would like to get updated.

**Comment by alepauly** [comment]
> 
> @alepauly, can you tell me who could review this change? It is something we would like to get updated.

Sorry for the delay @MichaelSimons, and thanks for the PR! I'm the bottleneck here trying to catch up. Added @bishal-pdMSFT who can help with a review.

**Comment by bishal-pdMSFT** [comment]
@hiyadav can you please review this.

**Comment by alepauly** [comment]
Thanks @MichaelSimons I'm running an internal build that does some checks and will merge after that succeeds.

**Comment by MichaelSimons** [comment]
@alepauly - any update on the verification?

**Comment by ShubhamGargCTH** [comment]
@alepauly @MichaelSimons @bishal-pdMSFT do we know when this PR will be merged ?

**Comment by alepauly** [comment]
@MichaelSimons sorry for the delay, the verification build failed and I forgot to follow up. See relevant part of the logs [here](https://gist.github.com/alepauly/4a3795632b9beb7b01a29d020b6e3f1a). I looked briefly and I couldn't tell why it would fail. @bishal-pdMSFT @hiyadav can you please look at why it's failing? 

as an aside, we're working on getting those builds as checks in these PRs, thanks for your patience!

**Comment by SimonCropp** [comment]
anything i can help with here?

**Comment by estensen** [comment]
@alepauly Is there an ETA for this?

File: images/linux/scripts/installers/1604/dotnetcore-sdk.sh [hunk 1]
```diff
@@ -7,7 +7,7 @@
 source $HELPER_SCRIPTS/apt.sh
 source $HELPER_SCRIPTS/document.sh
 
-LATEST_DOTNET_PACKAGE=dotnet-sdk-3.0
+LATEST_DOTNET_PACKAGE=dotnet-sdk-3.1
```
Comment (hiyadav) [inline images/linux/scripts/installers/1604/dotnetcore-sdk.sh]: LATEST_DOTNET_PACKAGES=("dotnet-sdk-3.0" "dotnet-sdk-3.1")

LSB_RELEASE=$(lsb_release -rs)

mksamples()
{
    sdk=$1
    sample=$2
    mkdir "$sdk"
    cd "$sdk" || exit
    set -e
    dotnet help
    dotnet new globaljson --sdk-version "$sdk"
    dotnet new "$sample"
    dotnet restore
    dotnet build
    set +e
    cd .. || exit
    rm -rf "$sdk"
}

set -e

for latest_package in ${LATEST_DOTNET_PACKAGES[@]}; do
    echo "Determing if .NET Core ($latest_package) is installed"
    if ! IsInstalled $latest_package; then
        echo "Could not find .NET Core ($latest_package), installing..."
        curl https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > microsoft.gpg
        mv microsoft.gpg /etc/apt/trusted.gpg.d/microsoft.gpg
        sh -c 'echo "deb [arch=amd64] https://packages.microsoft.com/repos/microsoft-ubuntu-bionic-prod bionic main" > /etc/apt/sources.list.d/dotnetdev.list'
        apt-get install apt-transport-https
        apt-get update
        apt-get install $latest_package -y
    else
        echo ".NET Core ($latest_package) is already installed"
    fi
done

**Comment by hiyadav** [comment]
Hey @MichaelSimons 
The installation script is failing for this change, as in this script we are now copying the `Templates` folder present at the root of extracted tars'. 
This was not a requirement till SDK version 2.2 as the templates folder used to be inside the sdk folder. But starting from 3.0 the templates have been moved out. 
Also it did not fail when we made the change to install 3.0 because one of the SDK versions of 3.0 was being installed from microsoft registries.

As a workaround to merge this PR, you can make the changes in this comment:https://github.com/actions/virtual-environments/pull/164#discussion_r371825065

Later, we will update and investigate how to handle multiple sdk/runtimes for versions 3.0 and above

**Comment by MichaelSimons** [comment]
@hiyadav, I can make that change.  Since the PR's creation, .NET 2.2. has reached its EOL.  Would it be an appropriate time to remove 2.2 (e.g. https://github.com/actions/virtual-environments/pull/164/files#diff-f26b5b40e3d8e91ff8143fb8b076d114R47)?

**Comment by alepauly** [comment]
> Would it be an appropriate time to remove 2.2

@MichaelSimons I think we should leave 2.2 for now since it's a breaking change (even though it's EOL'd some people have builds still using it). Thank you!

**Comment by hiyadav** [comment]
@MichaelSimons for now, keep the 2.2 version. Removing it will cause lot of failures.
We will stage its removal by first showing some warning messages in our tasks. Then later we can remove it

**Comment by MichaelSimons** [comment]
That's what I was suspecting but wanted to check.

**Comment by MichaelSimons** [comment]
I made the suggested updates.  The METADATA_FILE generated from running these changes looks like the following:

- .NET Core SDK:
  - 3.1.101
  - 3.1.100
  - 3.0.102
  - 3.0.101
  - 3.0.100
  - 2.2.402
  - 2.2.401
  - 2.2.301
  - 2.2.300
  - 2.2.207
  - 2.2.206
  - 2.2.205
  - 2.2.204
  - 2.2.203
  - 2.2.202
  - 2.2.110
  - 2.2.109
  - 2.2.108
  - 2.2.107
  - 2.2.106
  - 2.2.105
  - 2.2.104
  - 2.2.103
  - 2.2.102
  - 2.2.101
  - 2.2.100
  - 2.1.803
  - 2.1.802
  - 2.1.801
  - 2.1.701
  - 2.1.700
  - 2.1.608
  - 2.1.607
  - 2.1.606
  - 2.1.605
  - 2.1.604
  - 2.1.603
  - 2.1.602
  - 2.1.511
  - 2.1.510
  - 2.1.509
  - 2.1.508
  - 2.1.507
  - 2.1.506
  - 2.1.505
  - 2.1.504
  - 2.1.503
  - 2.1.502
  - 2.1.500
  - 2.1.403
  - 2.1.402
  - 2.1.401
  - 2.1.302
  - 2.1.301
  - 2.1.300

**Comment by estensen** [comment]
How long will packages that has reached end-of-life be kept? Is there a policy or just a gut-feeling?

**Comment by alepauly** [comment]
thanks @MichaelSimons, I'm running a verification build now. @hiyadav can you do a final review of this please?

**Comment by alepauly** [comment]
> How long will packages that has reached end-of-life be kept? Is there a policy or just a gut-feeling?

No official policy, right now we'll cleanup as necessary and possible. If we're removing something that would break existing builds we will announce it prior to doing it so people will have to react. This announcement is normally through a pinned issue in the repo.

**Comment by hiyadav** [comment]
@alepauly we are good to merge this PR, the script was successfully able to install all required .NET Core versions and run tests on them !  Can you merge this change as I'm yet to get permissions for this repo... 

@MichaelSimons thanks for bringing the feature alive :)

**Comment by SimonCropp** [comment]
thanks for the effort peoples. from this process, is there anything that needs to be extracted into a "how we update .netcore" document? or does such a doc already exist?

**Comment by dtekchandani** [comment]
@MichaelSimons @alepauly When will these changes be available in release branches?

**Comment by miketimofeev** [comment]
@dtekchandani these are supposed to be available in the next 1-2 weeks

**Comment by alepauly** [comment]
@dtekchandani what @miketimofeev said. Once we start deploying it, the Releases in this repo will show as pre-release. Becomes official release once it reaches all runners (takes about a week after that).

**Comment by dkmiller** [comment]
Is there any status update on deployment of this change? Ubuntu 'latest' images still only have .NET Core 3.0.

**Comment by maxim-lobanov** [comment]
@dkmiller , 
PR #164 was merged recently, Changes will be delivered in 1 week.

**Comment by dkmiller** [comment]
Ubuntu 'latest' images are now picking up .NET Core 3.1. Thanks!

**Comment by nietras** [comment]
> Later, we will update and investigate how to handle multiple sdk/runtimes for versions 3.0 and above

@hiyadav how can I have multiple dotnet core sdks in GitHub Actions? I would like to have both 3.1 and 5.0 (preview), this seems to work on Windows because 3.1 perhaps is part of default available on windows image, but on MacOS and Linux 3.1 is not available even though I set it up with setup-dotnet action.

**Pull request merged**